table/tables/partition.go

```diff
@@ -263,6 +267,55 @@ type PartitionExpr struct {
 	*ForListPruning
 }
 
+// GetPartColumnsForKeyPartition is used to get partition columns for key partition table
+func (pe *PartitionExpr) GetPartColumnsForKeyPartition(columns []*expression.Column) ([]*expression.Column, []int) {
+	schema := expression.NewSchema(columns...)
+	partCols := make([]*expression.Column, len(pe.ColumnOffset))
+	colLen := make([]int, 0, len(pe.ColumnOffset))
+	for i, offset := range pe.ColumnOffset {
+		partCols[i] = schema.Columns[offset]
+		partCols[i].Index = i
+		colLen = append(colLen, partCols[i].RetType.GetFlen())
+	}
+	return partCols, colLen
+}
+
+// LocateKeyPartitionWithSPC is used to locate the destination partition for key
+// partition table has single partition column(SPC). It's called in FastPlan process.
+func (pe *PartitionExpr) LocateKeyPartitionWithSPC(pi *model.PartitionInfo,
+	r []types.Datum) (int, error) {
+	col := &expression.Column{}
+	*col = *pe.KeyPartCols[0]
+	col.Index = 0
+	return pe.LocateKeyPartition(pi, []*expression.Column{col}, r)
+}
+
+// LocateKeyPartition is the common interface used to locate the destination partition
+func (pe *PartitionExpr) LocateKeyPartition(pi *model.PartitionInfo,
+	cols []*expression.Column, r []types.Datum) (int, error) {
+	nr1 := uint64(1)
+	nr2 := uint64(4)
+	for _, col := range cols {
+		if r[col.Index].Kind() == types.KindNull {
+			nr1 ^= (nr1 << 1) | 1
+			continue
+		}
+		data, err := r[col.Index].ToBytes()
+		if err != nil {
+			return 0, err
+		}
+		nr1, nr2 = types.CalcBytesHash(data, nr1, nr2)
+	}
+
+	partID := nr1 & pi.HashMask
+	if partID > pi.Num {
+		newmask := ((pi.HashMask + 1) >> 1) - 1
+		partID = nr1 & newmask
+	}
```

Comment:
This looks like LINEAR KEY and not KEY partitioning? I would assume non LINEAR would just use modulus?

Reply:
Yes, it's linear, I'll fix it.